src/main/java/org/broadinstitute/hellbender/engine/ReadWindowWalker.java

```diff
@@ -0,0 +1,198 @@
+package org.broadinstitute.hellbender.engine;
+
+
+import com.google.common.annotations.VisibleForTesting;
+import htsjdk.samtools.SAMSequenceDictionary;
+import org.broadinstitute.hellbender.cmdline.Argument;
+import org.broadinstitute.hellbender.engine.filters.CountingReadFilter;
+import org.broadinstitute.hellbender.engine.filters.ReadFilterLibrary;
+import org.broadinstitute.hellbender.engine.filters.WellformedReadFilter;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A ReadWindowWalker is a tool that processes entire regions/windows of reads at a time. These windows may or may not be overlapping.
+ *
+ * Reads in each window are lazily fetched rather than loaded all at once (though tools may choose to load all the reads in a window
+ * at once).
+ *
+ * {@link #windowSize} controls the maximum size of each window. Large intervals are sharded into regions of up to this size,
+ * while intervals smaller than this size are not subdivided.
+ *
+ * {@link #windowPadding} controls the number of bases of extra context to fetch on either side of the window. Overlapping reads
+ * are fetched for the padded regions, but they are marked in the {@link ReadWindow} objects as not being within the window proper.
+ *
+ * If multiple sources of reads are specified, they are merged together into a single sorted stream of reads.
+ *
+ * ReadWindowWalker authors must implement the {@link #apply}, {@link #defaultWindowSize}, and {@link #defaultWindowPadding} methods,
+ * and may optionally implement {@link #onTraversalStart} and/or {@link @onTraversalDone}.
+ */
+public abstract class ReadWindowWalker extends GATKTool {
+
+    @Argument(fullName="windowSize", shortName="windowSize", doc = "Maximum size of each window, in bases", optional = true)
+    protected int windowSize = defaultWindowSize();
+
+    @Argument(fullName="windowPadding", shortName="windowPadding", doc = "Each window has this many bases of extra context on each side", optional = true)
+    protected int windowPadding = defaultWindowPadding();
+
+    @Argument(fullName = "disable_all_read_filters", shortName = "f", doc = "Disable all read filters", common = false, optional = true)
+    public boolean disableAllReadFilters = false;
+
+    private List<ReadWindow> windows;
+
+    @Override
+    public boolean requiresReads() {
+        return true;
+    }
+
+    /**
+     * @return Default value for the {@link #windowSize} parameter, if none is provided on the command line
+     */
+    protected abstract int defaultWindowSize();
```

Comment:
i find these 2 methods (`defaultWindowSize` and `defaultWindowPadding`) misplaced here. If this abstract class does not use them and only the HC subclass does, then they belong in the subclass and not here.